cmd/test-server/networking.go

```diff
@@ -0,0 +1,366 @@
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/go-ping/ping"
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcap"
+	"golang.org/x/net/ipv4"
+)
+
+const (
+	ICMPCount            = 5
+	ICMPTimeout          = time.Second * 10
+	beginTTLValue        = 5
+	MaxTTLHops           = 32
+	stringToSend         = "test string tcp"
+	tracerouteHopTimeout = time.Second * 10
+	snaplen              = 65536
+	promisc              = true
+)
+
+var (
+	buffer  gopacket.SerializeBuffer
+	options = gopacket.SerializeOptions{
+		ComputeChecksums: true,
+		FixLengths:       true,
+	}
+	timerPerHopPerUUID = make(map[string]time.Time)
+	deviceName         string
+	currTTLIndicator   = make(map[string]int)
+	icmpPktError       = errors.New("IP header unavailable")
+)
+
+type RtItem struct {
+	IP        string
+	PktSent   int
+	PktRecv   int
+	PktLoss   float64
+	MinRtt    float64
+	AvgRtt    float64
+	MaxRtt    float64
+	StdDevRtt float64
+}
+
+type Results struct {
+	UUID        string
+	IPaddr      string
+	Timestamp   string
+	IcmpPing    []RtItem
+	AvgIcmpStat float64
+}
+
+type SentPacketData struct {
+	HopIPId     uint16
+	HopSentTime time.Time
+}
+
+type HopRTT struct {
+	IP  net.IP
+	RTT float64
+}
+
+type TracerouteResults struct {
+	UUID      string
+	Timestamp string
+	HopData   map[int]HopRTT
+}
+
+// IcmpPinger sends ICMP pings and returns statistics
+func IcmpPinger(ip string) RtItem {
+	pinger, err := ping.NewPinger(ip)
+	if err != nil {
+		panic(err)
+	}
+	pinger.Count = ICMPCount
+	pinger.Timeout = ICMPTimeout
+	err = pinger.Run() // Blocks until finished.
+	if err != nil {
+		panic(err)
+	}
+	stat := pinger.Statistics()
+	icmp := RtItem{ip, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt),
+		fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)}
+	return icmp
+}
+
+// getMeanIcmpRTT gets Avg RTT from all successful ICMP measurements, to display on webpage
+func getMeanIcmpRTT(icmp []RtItem) float64 {
+	var sum float64 = 0
+	var len float64 = 0
+	for _, x := range icmp {
+		if x.AvgRtt == 0 {
+			continue
+		}
+		sum += x.AvgRtt
+		len += 1
+	}
+	var avg float64 = sum / len
+	if math.IsNaN(avg) {
+		return 0.0
+	}
+	return avg
+}
+
+// start0trace reaches the underlying connection and sets up necessary pcap handles
+// and implements the 0trace method of sending TTL-limited probes on an existing TCP connection
+func start0trace(uuid string, netConn net.Conn) map[int]HopRTT {
+	clientIPstr := netConn.RemoteAddr().String()
+	clientIP, clPort, _ := net.SplitHostPort(clientIPstr)
+	clientPort, _ := strconv.Atoi(clPort)
+
+	handle, err := pcap.OpenLive(deviceName, snaplen, promisc, time.Second)
+	if err != nil {
+		ErrLogger.Println("Handle error:", err)
+	}
+
+	if err = handle.SetBPFFilter(fmt.Sprintf("(tcp and port %d and host %s) or icmp", clientPort, clientIP)); err != nil {
+		log.Fatal(err)
```

Comment:
Should these `log.Fatal` errors be logged into ErrLogger instead? Otherwise it writes to standard error.

Reply:
(Or are Fatal logs just logged to stderr generally?)

Reply:
That's fair. Let's use the ErrLogger for consistency. Fixed in 6471c18c38447c7bd56aa218fe6fa4f42c38cfc0. And yes, the log module's standard logger writes to stderr.